amads/core/vectors_sets.py

```diff
@@ -1,190 +1,249 @@
     >>> resulting_set
     (1, 1, 1, 2, 2, 3)
 
-    >>> roundtrip = set_to_vector(resulting_set, max_index=6)
+    >>> roundtrip = multiset_to_vector(resulting_set, max_index=6)
     >>> roundtrip
     (0, 3, 2, 1, 0, 0, 0)
 
     >>> roundtrip == test_vector
     True
 
     """
-    return tuple(i for i in range(len(vector)) for _ in range(vector[i]))
+    multiset = []
+    for i in range(len(vector)):
+        for _ in range(vector[i]):
+            multiset.append(i)
+
+    return tuple(multiset)
 
 
-# Arithmetic operations: Addition/subtraction, multiplication,division
+# Arithmetic operations: Addition/subtraction, multiplication, division
 
 
 def apply_constant(
-    set_or_vector: list[int] | tuple[int], constant: float, modulo: int | None = None
-) -> list | tuple:
+    set_or_vector: Iterable,
+    constant: float,
+    modulo: Union[int, None] = None,
+) -> Union[tuple, list, set]:
     """
     Apply a constant value to a set or vector.
 
     Parameters
     ----------
-    set_or_vector: Any list or tuple representing a set or vector.
-    constant: The constant may be an int or a float and positive or negative.
-    modulo: Int or `None`. If int, return the values modulo this value (e.g., 12 for pitch class sets)
+    set_or_vector:
+        An iterable representing a set or vector.
+    constant:
+        The constant may be an int or a float and positive or negative.
+    modulo:
+        An optional integer.
+        If provided, the result will be taken modulo this value.
 
     Returns
     -------
-    list | tuple: The set or vector with constant applied.
+    tuple, list, or set:
+        The set or vector with the constant applied.
+        This return type matches the input type.
 
     Examples
     --------
-
-    >>> start = [0, 1, 2, 3, 11]
+    >>> start = (0, 1, 2, 3, 11)
     >>> more = apply_constant(start, 1)
     >>> more
-    [1, 2, 3, 4, 12]
+    (1, 2, 3, 4, 12)
 
     >>> less = apply_constant(more, -1)
     >>> less
-    [0, 1, 2, 3, 11]
+    (0, 1, 2, 3, 11)
 
     >>> start == less
     True
 
     >>> more = apply_constant(start, 1, modulo=12)
     >>> more
-    [1, 2, 3, 4, 0]
+    (1, 2, 3, 4, 0)
+
+    >>> as_list = [0, 1, 2, 3, 11]
+    >>> more_list = apply_constant(as_list, 1)
+    >>> more_list
+    [1, 2, 3, 4, 12]
+
+    >>> as_set = {0, 1, 2, 3, 11}
+    >>> more_set = apply_constant(as_set, 1)
+    >>> more_set
+    {1, 2, 3, 4, 12}
     """
     result = [x + constant for x in set_or_vector]
     if modulo is not None:
         result = [x % modulo for x in result]
-    return result
+
+    if isinstance(set_or_vector, tuple):
+        return tuple(result)
+    elif isinstance(set_or_vector, list):
+        return result
+    elif isinstance(set_or_vector, set):
+        return set(result)
+    else:
+        return result
 
 
-def scalar_multiply(input_list: list, scale_factor: int = 2) -> list:
+def scalar_multiply(input_tuple: tuple, scale_factor: int = 2) -> tuple:
```

Comment:
There is something redundant (here and elsewhere about specifying the object type both in its name and in its type hint. I think `input` would be fine here (or `object`, or `x`)!